tests/scripts/thread-cert/test_radio_filter.py

```diff
@@ -0,0 +1,143 @@
+#     derived from this software without specific prior written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+#  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+#  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+#  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+#  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+#
+
+import ipaddress
+import unittest
+
+import config
+import command
+import thread_cert
+
+# Test description:
+#
+# This test verifies the radio filter mechanism.
+#
+# Topology:
+#
+#   Leader -- Router
+#     |
+#     |
+#   Child
+#
+#
+
+LEADER = 1
+ROUTER = 2
+SED = 3
+
+WAIT_TIME = 5
+
+
+class RadioFilter(thread_cert.TestCase):
+    USE_MESSAGE_FACTORY = False
+    SUPPORT_NCP = False
+
+    TOPOLOGY = {
+        LEADER: {
+            'mode': 'rdn',
+            'allowlist': [ROUTER, SED]
+        },
+        ROUTER: {
+            'mode': 'rdn',
+            'allowlist': [LEADER]
+        },
+        SED: {
+            'is_mtd': True,
+            'mode': '-',
+            'timeout': config.DEFAULT_CHILD_TIMEOUT,
+            'allowlist': [LEADER]
+        },
+    }
+
+    def test(self):
+        leader = self.nodes[LEADER]
+        router = self.nodes[ROUTER]
+        sed = self.nodes[SED]
+
+        nodes = [leader, router, sed]
+
+        leader.start()
+        self.simulator.go(WAIT_TIME)
+        self.assertEqual(leader.get_state(), 'leader')
+
+        router.start()
+        self.simulator.go(WAIT_TIME)
+        self.assertEqual(router.get_state(), 'router')
+
+        sed.start()
+        sed.set_pollperiod(40)
+        self.simulator.go(WAIT_TIME)
+        self.assertEqual(sed.get_state(), 'child')
+
+        # Validate initial state of `radiofilter` upon start
+        for node in nodes:
+            self.assertFalse(node.radiofilter_is_enabled())
+
+        leader_mleid = leader.get_mleid()
+        router_mleid = router.get_mleid()
+        sed_mleid = sed.get_mleid()
+
+        # Validate connections by pinging from leader, SED, and router
+        self.assertTrue(leader.ping(router_mleid))
+        self.assertTrue(sed.ping(leader_mleid))
+        self.assertTrue(router.ping(leader_mleid))
+
+        # Valid behavior when `radiofilter` is enabled on router
+        router.radiofilter_enable()
+        self.assertTrue(router.radiofilter_is_enabled())
+        self.assertFalse(leader.ping(router_mleid))
```

Comment:
test energy scan during radio filter enabled?

Reply:
I considered it but the thing was we did not seem to have `energy scan` support in `node.py`. I ended up adding in new push (I was initially tempted to not even add a test for `radiofilter` as was hoping to spend less time on this).
@simonlingoogle , you reached out mentioning you were busy and this was a feature on your plate. I offered to help and added it quickly. You owe me one man 😄

Reply:
Thanks.